cvat/apps/dataset_manager/bindings.py

```diff
@@ -2151,11 +2151,87 @@ def reduce_fn(acc, v):
                 raise CvatImportError("Image {}: can't import annotation "
                     "#{} ({}): {}".format(item.id, idx, ann.type.name, e)) from e
 
-    for track in tracks.values():
+    stop_frame = int(instance_data.meta[instance_data.META_FIELD]['stop_frame'])
+    for track_id, track in tracks.items():
+        # skip tracks with no shapes and elements
+        if len(track['shapes']) == 0 and len(track['elements']) == 0:
+            continue
+
+        track['shapes'].sort(key=lambda t: t.frame)
+        prev_shape = track['shapes'][0]
+        # add keyframe to the first non-outside shape
+        if not prev_shape.outside:
+            prev_shape = prev_shape._replace(keyframe=True)
+        new_shapes = []
+
+        # avoid skipping the first frame
+        if len(track['shapes']) > 1:
+            for shape in track['shapes'][1:]:
+                if prev_shape.keyframe:
+                    new_shapes.append(prev_shape)
+                else:
+                    has_skip = instance_data.frame_step < shape.frame - prev_shape.frame
+                    if has_skip and not prev_shape.outside:
+                        prev_shape = prev_shape._replace(outside=True, keyframe=True,
+                                frame=prev_shape.frame + instance_data.frame_step)
+                        new_shapes.append(prev_shape)
+                prev_shape = shape
+
+            if prev_shape.outside or prev_shape.keyframe:
+                new_shapes.append(prev_shape)
+            elif not prev_shape.outside and prev_shape.frame + instance_data.frame_step <= stop_frame:
+                prev_shape = prev_shape._replace(outside=True, keyframe=True,
+                        frame=prev_shape.frame + instance_data.frame_step)
+                new_shapes.append(prev_shape)
+
+        else:
+            if prev_shape.keyframe:
+                new_shapes.append(prev_shape)
```

Comment:
```suggestion
new_shapes = []
        prev_shape = None
        # infer and keep only the keyframe shapes
        for shape in track['shapes']:
            cur_is_visible = shape and not shape.outside
            if not cur_is_visible:
                continue

            prev_is_visible = prev_shape and not prev_shape.outside
            if not prev_is_visible or (
                has_gap := prev_shape.frame + instance_data.frame_step < shape.frame
            ):
                if has_gap:
                    prev_shape = prev_shape._replace(outside=True, keyframe=True,
                        frame=prev_shape.frame + instance_data.frame_step)
                    new_shapes.append(prev_shape)

                shape = shape._replace(keyframe=True)
                new_shapes.append(shape)
                prev_shape = shape

        if prev_shape and not prev_shape.outside and (
            prev_shape.frame + instance_data.frame_step <= stop_frame
            # has a gap before the current instance segment end
        ):
            prev_shape = prev_shape._replace(outside=True, keyframe=True,
                frame=prev_shape.frame + instance_data.frame_step)
            new_shapes.append(prev_shape)
```

I suggest that we use something like this. Note that this needs to be replicated for track elements as well. Maybe it's worth be extracted as a function to avoid duplicating code.